frontend/javascripts/oxalis/model/bucket_data_handling/data_cube.js

```diff
@@ -367,6 +377,120 @@ class DataCube {
     }
   }
 
+  floodFill(
+    initialVoxel: Vector3,
+    cellId: number,
+    get3DAddress: (Vector2, Vector3) => Vector3,
+    get2DAddress: Vector3 => Vector2,
+    dimensionsToIterateOver: Vector2,
+    viewportBoundings: BoundingBoxType,
+    zoomStep: number = 0,
+  ) {
+    // This flood-fill algorithm works in two nested levels and uses a list of buckets to flood fill.
+    // On the inner level a bucket is flood-filled  and if the iteration of the buckets data
+    // reaches an neighbour bucket, this bucket is added to this list of buckets to flood fill.
+    // The outer level simply iterates over all  buckets in the list and triggers the bucket-wise flood fill.
+    //
+    // Note: It is possible that a bucket is multiple times added to the address. This is intended
+    // because a border of the "neighbour volume shape" might leave the neighbour bucket and enter is somewhere else.
```

Comment:
Infinite loops should be avoided because of the following check before marking anything changes to the currentBucket:
```
if (bucketData[initialVoxelIndex] !== sourceCellId) {
    // Ignoring neighbour buckets whose cellId at the initial voxel does not match the source cell id.
    continue;
}
```
This prevents to start a flood fill at a border-voxel of the bucket, that is already label or that has a different id than the id that should be overwritten.
 As a bucket has only a limit of border-voxel and the border-voxel will always be marked after this check (see below), an infinite loop should be impossible

Note:
Ok there is one possibility for an error if only a single voxel at the border of a bucket should be flood filled. Then this voxel has no neighbours and thus will not be marked. I just fixed this.

But I do not think that this was what caused the infinite loop for you @philippotto.

Reply:
> Ok there is one possibility for an error if only a single voxel at the border of a bucket should be flood filled. Then this voxel has no neighbours and thus will not be marked. I just fixed this.

Did you already push it? I think, I have a way to reproduce the problem. Paint a cell blue and than initiate the flood fill with the same cell id at a voxel where the cell is already blue. Probably, nothing should happen, but appears to run into an infinite loop for me. Can you confirm?

Reply:
Nope, I am still debugging this. 
> Paint a cell blue and then initiate the flood fill with the same cell id at a voxel where the cell is already blue. Probably, nothing should happen but appears to run into an infinite loop for me. Can you confirm?

I already thought of that but did not test this out. Looking at the code: I just thought of that case but did not implement a  fix for this. I'll push the fix, once I fixed another bug I found resulting in an infinite loop.